### PR TITLE
Implement approx_median in BQ

### LINKED ISSUE #1604: Implementing approx_quantile, approx_median for BQ
approx_quantile is BQ function that does not exist in IBIS
approx_median is an IBIS function that is not implemented for BQ

Can we implement the ibis function approx_median on BQ using the approx_quantile function?

### PR DESCRIPTION
**Pull request by missing-semicolon**
Closes #1604

### CODE REVIEW

File: ibis/bigquery/compiler.py [hunk 1]
```diff
@@ -586,6 +586,19 @@ def compiles_floor(t, e):
...
 
+@compiles(ops.CMSMedian)
+def compiles_approx(translator, expr):
+    arg = expr.op().args[0]
+    where = expr.op().args[1]
+
+    if where is not None:
+        arg = where.ifelse(arg, ibis.NA)
+
+    return 'APPROX_QUANTILES({}, 2)[OFFSET(1)]'.format(
```
Comment (missing-semicolon) [inline ibis/bigquery/compiler.py]: `APPROX_QUANTILES` returns an array. By setting the number of quantiles to 2 it returns an array of  [min, median, max], so the offset is required to pull out the median.

**Comment by missing-semicolon** [comment]
It occurs to me that I do not know the true medians in order to implement a client test. Plus the responses are approximate. Any thoughts?

File: ibis/bigquery/compiler.py (hunk 1 above)
Comment (cpcloud) [reply ibis/bigquery/compiler.py]: Is there any reason to use [`SAFE_OFFSET`](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#safe_offset-and-safe_ordinal) here?

File: ibis/bigquery/compiler.py [hunk 2]
```diff
@@ -586,6 +586,19 @@ def compiles_floor(t, e):
     return 'CAST(FLOOR({}) AS {})'.format(t.translate(arg), bigquery_type)
 
 
+@compiles(ops.CMSMedian)
+def compiles_approx(translator, expr):
+    arg = expr.op().args[0]
+    where = expr.op().args[1]
```
Comment (cpcloud) [inline ibis/bigquery/compiler.py]: Let's just pull out `expr.op()` into a variable and use the `arg` and `where` properties.

**Comment by cpcloud** [comment]
The conda build failures are unrelated to your changes here.

**Comment by missing-semicolon** [comment]
Funny you said that, @cpcloud I was just thinking about computing in pandas and set up a client test for that. Looks like it's failing but I'm not sure why. `approx_median` is returning a 7 for the `alltypes.month` column when the actual one appears to be 6.

**Comment by missing-semicolon** [comment]
Re-upping this comment about `OFFSET` vs `SAFE_OFFSET` since it's hidden by the subsequent code-push:

> There shouldn't be an instance where the offset is out of bounds given that we've hard-coded 2 as the number of quantiles to create. That said, we would want this to fail noisily if it were an issue given that we're trying to calculate the median explicitly.

**Comment by missing-semicolon** [comment]
I got the client_test to pass  🎉

**Comment by cpcloud** [comment]
> There shouldn't be an instance where the offset is out of bounds given that we've hard-coded 2 as the number of quantiles to create. That said, we would want this to fail noisily if it were an issue given that we're trying to calculate the median explicitly

What if the table has no rows?

**Comment by cpcloud** [comment]
>> There shouldn't be an instance where the offset is out of bounds given that we've hard-coded 2 as the number of quantiles to create. That said, we would want this to fail noisily if it were an issue given that we're trying to calculate the median explicitly

> What if the table has no rows?

If there are no rows, an expression like `approx_quantiles(col, 2)[offset(1)]` will never be evaluated, and with 1 or more rows there will always be 3 elements, so using `OFFSET(1)` is fine.

**Comment by cpcloud** [comment]
@missing-semicolon can you rebase?

**Comment by cpcloud** [comment]
@missing-semicolon So the semantic of the `approx_median` function you've implemented raises a few questions/issues:

1. This implementation differs significantly from other implementations of `median` in that it doesn't return a continuous value for the (approximate) median.
1. The BigQuery [`percentile_cont` function](https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators#percentile_cont) appears to give that value that we want, except that it can only be used as an analytic function in a window operation. Bummer.
1. I'm not opposed to including this as is, with some additional documentation stating that the median returned here is coming from discrete quantiles and will thus give a potentially different answer than if you used a different system.

**Comment by cpcloud** [comment]
@missing-semicolon Can you rebase? I'll merge afterwards.

**Comment by missing-semicolon** [comment]
Just rebased 🚀

**Comment by cpcloud** [comment]
thanks @missing-semicolon !

**Pull request closed without merging**